File: server/utils/mpesa_bulk.py
"""Bulk sources of M-PESA history.

A browser cannot read a phone's SMS inbox, so "import everything at once" means
handing Kipato a file. Three shapes cover what workers actually have:

* **SMS Backup & Restore XML** — the usual Android SMS export.
* **M-PESA statement CSV** — the Receipt/Completion Time/Paid In export.
* **Plain text** — a long paste, or a .txt dump of messages.

Everything funnels into `parse_mpesa_text` so one set of rules decides what
counts as income, no matter where the text came from.
"""

import csv
import io
import re
import xml.etree.ElementTree as ElementTree
from datetime import datetime
from typing import List, Optional

from server.utils.mpesa_parser import MAX_AMOUNT, parse_mpesa_text

XML = "xml"
CSV_STATEMENT = "csv"
TEXT = "text"

# Column names in the M-PESA statement export, lowercased.
_RECEIPT_COLUMNS = ("receipt no.", "receipt no", "receipt", "transaction id")
_DATE_COLUMNS = ("completion time", "date", "date & time", "completion date")
_PAID_IN_COLUMNS = ("paid in", "paid_in", "credit", "amount received")
_DETAIL_COLUMNS = ("details", "description", "narrative")


def detect_format(raw_text: str) -> str:
    head = raw_text.lstrip()[:2000]
    if head.startswith("<?xml") or "<smses" in head or "<sms " in head:
        return XML
    if _csv_header(raw_text) is not None:
        return CSV_STATEMENT
    return TEXT


def parse_bulk(raw_text: str) -> List[dict]:
    """Parse any supported bulk format into the usual parsed-entry dicts."""
    if not raw_text:
        return []

    fmt = detect_format(raw_text)
    if fmt == XML:
        return _dedupe(_parse_sms_backup_xml(raw_text))
    if fmt == CSV_STATEMENT:
        return _dedupe(_parse_statement_csv(raw_text))
    return parse_mpesa_text(raw_text)


def _dedupe(entries: List[dict]) -> List[dict]:
    seen = set()
    unique = []
    for entry in entries:
        code = entry.get("code")
        if code:
            if code in seen:
                continue
            seen.add(code)
        unique.append(entry)
    return unique


def _parse_sms_backup_xml(raw_text: str) -> List[dict]:
    try:
        root = ElementTree.fromstring(raw_text)
    except ElementTree.ParseError:
        return []

    entries: List[dict] = []
    for element in root.iter():
        if element.tag.lower() not in ("sms", "mms"):
            continue

        address = (element.get("address") or "").upper()
        body = element.get("body") or ""
        if not body:
            continue
        # Keep messages from M-PESA. An unnamed sender still gets a look, since
        # the parser itself rejects anything that is not a confirmation.
        if address and "MPESA" not in address.replace("-", "").replace(" ", ""):
            if "M-PESA" not in address:
                continue
        entries.extend(parse_mpesa_text(body))

    return entries


def _csv_header(raw_text: str) -> Optional[List[str]]:
    try:
        reader = csv.reader(io.StringIO(raw_text))
        for row in reader:
            lowered = [cell.strip().lower() for cell in row]
            has_receipt = any(cell in _RECEIPT_COLUMNS for cell in lowered)
            has_paid_in = any(cell in _PAID_IN_COLUMNS for cell in lowered)
            if has_receipt and has_paid_in:
                return lowered
            # Statements carry a few preamble rows before the real header.
            if reader.line_num > 25:
                break
    except (csv.Error, UnicodeDecodeError):
        return None
    return None


def _column_index(header: List[str], names) -> Optional[int]:
    for index, cell in enumerate(header):
        if cell in names:
            return index
    return None


def _parse_statement_csv(raw_text: str) -> List[dict]:
    header = _csv_header(raw_text)
    if header is None:
        return []

    receipt_at = _column_index(header, _RECEIPT_COLUMNS)
    date_at = _column_index(header, _DATE_COLUMNS)
    paid_in_at = _column_index(header, _PAID_IN_COLUMNS)
    detail_at = _column_index(header, _DETAIL_COLUMNS)
    if paid_in_at is None:
        return []

    entries: List[dict] = []
    reached_header = False
    for row in csv.reader(io.StringIO(raw_text)):
        lowered = [cell.strip().lower() for cell in row]
        if not reached_header:
            reached_header = lowered == header
            continue
        if len(row) <= paid_in_at:
            continue

        amount = _to_amount(row[paid_in_at])
        if amount is None:
            # No money in on this row: it is a payment, withdrawal or a total.
            continue

        entry_date = _to_date(row[date_at]) if date_at is not None and len(row) > date_at else None
        if entry_date is None:
            continue

        code = row[receipt_at].strip().upper() if receipt_at is not None and len(row) > receipt_at else None
        details = row[detail_at].strip() if detail_at is not None and len(row) > detail_at else ""

        entries.append({
            "code": code or None,
            "amount": amount,
            "date": entry_date,
            "sender": _sender_from_details(details),
            "raw": (", ".join(cell.strip() for cell in row))[:500],
        })

    return entries


def _to_amount(value: str) -> Optional[float]:
    cleaned = (value or "").replace(",", "").replace("Ksh", "").replace("KES", "").strip()
    if not cleaned:
        return None
    try:
        amount = float(cleaned)
    except ValueError:
        return None
    if amount <= 0 or amount > MAX_AMOUNT:
        return None
    return amount


_STATEMENT_DATE_FORMATS = (
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%d",
    "%d-%m-%Y %H:%M:%S",
    "%d/%m/%Y %H:%M:%S",
    "%d/%m/%Y %H:%M",
    "%d/%m/%Y",
    "%d-%m-%Y",
    "%d %b %Y",
)


def _to_date(value: str):
    cleaned = (value or "").strip()
    if not cleaned:
        return None
    for fmt in _STATEMENT_DATE_FORMATS:
        try:
            return datetime.strptime(cleaned, fmt).date()
        except ValueError:
            continue
    return None


def _sender_from_details(details: str) -> Optional[str]:
    if not details:
        return None
    # "Funds received from JOHN DOE 254712345678" / "Pay Bill from ..."
    match = re.search(r"from\s+([A-Za-z][A-Za-z.'\- ]{1,60})", details, re.IGNORECASE)
    if match:
        return " ".join(match.group(1).split()).strip(" .,-")[:120] or None
    return details[:120]

File: frontend/src/lib/mpesa-bulk.ts
import { parseMpesaText } from './mpesa-parser'
import type { ParsedMpesaEntry } from '../types'

/**
 * Bulk sources of M-PESA history.
 *
 * A browser cannot read a phone's SMS inbox, so importing everything at once
 * means handing Kipato a file. Three shapes cover what workers actually have:
 * an SMS Backup & Restore XML export, an M-PESA statement CSV, or a long paste.
 *
 * Mirrors `server/utils/mpesa_bulk.py`; change them together.
 */

export type BulkFormat = 'xml' | 'csv' | 'text'

const RECEIPT_COLUMNS = ['receipt no.', 'receipt no', 'receipt', 'transaction id']
const DATE_COLUMNS = ['completion time', 'date', 'date & time', 'completion date']
const PAID_IN_COLUMNS = ['paid in', 'paid_in', 'credit', 'amount received']
const DETAIL_COLUMNS = ['details', 'description', 'narrative']

const MAX_AMOUNT = 1_000_000

export function detectFormat(rawText: string): BulkFormat {
  const head = rawText.trimStart().slice(0, 2000)
  if (head.startsWith('<?xml') || head.includes('<smses') || head.includes('<sms ')) return 'xml'
  if (findCsvHeader(rawText) !== null) return 'csv'
  return 'text'
}

export function parseBulk(rawText: string): ParsedMpesaEntry[] {
  if (!rawText) return []

  switch (detectFormat(rawText)) {
    case 'xml':
      return dedupe(parseSmsBackupXml(rawText))
    case 'csv':
      return dedupe(parseStatementCsv(rawText))
    default:
      return parseMpesaText(rawText)
  }
}

function dedupe(entries: ParsedMpesaEntry[]): ParsedMpesaEntry[] {
  const seen = new Set<string>()
  const unique: ParsedMpesaEntry[] = []
  for (const entry of entries) {
    if (entry.code) {
      if (seen.has(entry.code)) continue
      seen.add(entry.code)
    }
    unique.push(entry)
  }
  return unique
}

function parseSmsBackupXml(rawText: string): ParsedMpesaEntry[] {
  let document: Document
  try {
    document = new DOMParser().parseFromString(rawText, 'application/xml')
  } catch {
    return []
  }
  if (document.querySelector('parsererror')) return []

  const entries: ParsedMpesaEntry[] = []
  for (const element of document.querySelectorAll('sms, mms')) {
    const body = element.getAttribute('body') ?? ''
    if (!body) continue

    const address = (element.getAttribute('address') ?? '').toUpperCase()
    // Keep messages from M-PESA. An unnamed sender still gets a look, since the
    // parser itself rejects anything that is not a confirmation.
    if (address && !address.replace(/[-\s]/g, '').includes('MPESA')) continue

    entries.push(...parseMpesaText(body))
  }
  return entries
}

/** Minimal RFC 4180 reader: enough for quoted "2,000.00" amounts. */
function parseCsvRows(rawText: string): string[][] {
  const rows: string[][] = []
  let row: string[] = []
  let field = ''
  let inQuotes = false

  for (let index = 0; index < rawText.length; index += 1) {
    const char = rawText[index]

    if (inQuotes) {
      if (char === '"') {
        if (rawText[index + 1] === '"') {
          field += '"'
          index += 1
        } else {
          inQuotes = false
        }
      } else {
        field += char
      }
      continue
    }

    if (char === '"') {
      inQuotes = true
    } else if (char === ',') {
      row.push(field)
      field = ''
    } else if (char === '\n' || char === '\r') {
      if (char === '\r' && rawText[index + 1] === '\n') index += 1
      row.push(field)
      rows.push(row)
      row = []
      field = ''
    } else {
      field += char
    }
  }

  if (field !== '' || row.length > 0) {
    row.push(field)
    rows.push(row)
  }
  return rows
}

function findCsvHeader(rawText: string): string[] | null {
  const rows = parseCsvRows(rawText)
  // Statements carry a few preamble rows before the real header.
  for (const row of rows.slice(0, 25)) {
    const lowered = row.map((cell) => cell.trim().toLowerCase())
    const hasReceipt = lowered.some((cell) => RECEIPT_COLUMNS.includes(cell))
    const hasPaidIn = lowered.some((cell) => PAID_IN_COLUMNS.includes(cell))
    if (hasReceipt && hasPaidIn) return lowered
  }
  return null
}

function columnIndex(header: string[], names: string[]): number {
  return header.findIndex((cell) => names.includes(cell))
}

function parseStatementCsv(rawText: string): ParsedMpesaEntry[] {
  const header = findCsvHeader(rawText)
  if (!header) return []

  const receiptAt = columnIndex(header, RECEIPT_COLUMNS)
  const dateAt = columnIndex(header, DATE_COLUMNS)
  const paidInAt = columnIndex(header, PAID_IN_COLUMNS)
  const detailAt = columnIndex(header, DETAIL_COLUMNS)
  if (paidInAt < 0) return []

  const entries: ParsedMpesaEntry[] = []
  let reachedHeader = false

  for (const row of parseCsvRows(rawText)) {
    const lowered = row.map((cell) => cell.trim().toLowerCase())
    if (!reachedHeader) {
      reachedHeader = lowered.length === header.length && lowered.every((cell, i) => cell === header[i])
      continue
    }
    if (row.length <= paidInAt) continue

    const amount = toAmount(row[paidInAt])
    // No money in on this row: it is a payment, withdrawal or a total.
    if (amount === null) continue

    const date = dateAt >= 0 && row.length > dateAt ? toIsoDate(row[dateAt]) : null
    if (!date) continue

    const code = receiptAt >= 0 && row.length > receiptAt ? row[receiptAt].trim().toUpperCase() : ''
    const details = detailAt >= 0 && row.length > detailAt ? row[detailAt].trim() : ''

    entries.push({
      code: code || null,
      amount,
      date,
      sender: senderFromDetails(details),
      rawMatch: row.map((cell) => cell.trim()).join(', ').slice(0, 500),
    })
  }

  return entries
}

function toAmount(value: string): number | null {
  const cleaned = (value ?? '').replace(/,/g, '').replace(/Ksh|KES/gi, '').trim()
  if (!cleaned) return null
  const amount = Number(cleaned)
  if (!Number.isFinite(amount) || amount <= 0 || amount > MAX_AMOUNT) return null
  return amount
}

function toIsoDate(value: string): string | null {
  const cleaned = (value ?? '').trim()
  if (!cleaned) return null

  const iso = /^(\d{4})-(\d{2})-(\d{2})/.exec(cleaned)
  if (iso) return `${iso[1]}-${iso[2]}-${iso[3]}`

  const dmy = /^(\d{1,2})[/-](\d{1,2})[/-](\d{4})/.exec(cleaned)
  if (dmy) {
    const day = dmy[1].padStart(2, '0')
    const month = dmy[2].padStart(2, '0')
    if (Number(month) < 1 || Number(month) > 12 || Number(day) < 1 || Number(day) > 31) return null
    return `${dmy[3]}-${month}-${day}`
  }
  return null
}

function senderFromDetails(details: string): string | null {
  if (!details) return null
  const match = /from\s+([A-Za-z][A-Za-z.'\- ]{1,60})/i.exec(details)
  if (match) {
    const sender = match[1].split(/\s+/).join(' ').replace(/^[.,\-\s]+|[.,\-\s]+$/g, '')
    return sender ? sender.slice(0, 120) : null
  }
  return details.slice(0, 120)
}
